### ISSUE TITLE
curio and tkinter

### ISSUE THREAD

**Issue by nameoftherose** [dabeaz/curio#111]
Thanks for curio and the youtube tutorials. They are great.
I have a number of sensors sending their measurements over the LAN. Using curio I collect all measurements.
I now need to integrate tkinter to the program.
I would be thankfull for any suggestions.
Thank you.

**Comment by nameoftherose** [comment]
[Neat discovery: How to combine asyncio and tkinter : Python - Reddit](https://www.google.gr/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwiSvr3k_q3QAhXFCBoKHQ9dAf0QFgggMAA&url=https%3A%2F%2Fwww.reddit.com%2Fr%2FPython%2Fcomments%2F33ecpl%2Fneat_discovery_how_to_combine_asyncio_and_tkinter%2F&usg=AFQjCNHGx5vWGlX0PLFj5HmkvqQQY0vwhw) shows a way to do this using the asyncio module. The concept translates in a straightforward way to **curio** with one exception: in order to spawn a new coroutine in response to a gui event, it uses the asyncio.async function (equivalently the asyncio.ensure_future). 
It seems there is no such functionality in curio. (curio.spawn is an async def function, cannot be called from a tkinter callback). 
Could such a function be added?  Or could you give some advice so I try to add it myself?
Thank you.

**Comment by dabeaz** [comment]
That is pretty interesting.  I have a solution! (I think).   The solution is to use a curio queue object.  Create a queue and a task that is solely dedicated to spawning new tasks::

     import curio
     newtask_queue = curio.Queue()

     async def creator_task():
              while True:
                      coro = await newtask_queue.get()
                      await curio.spawn(coro)

Launch this task when your curio application starts::

    async def main():
         await curio.spawn(creator_task())
         ...

Now, in your Tk callbacks, use the `newtask_queue.put()` method.   For example::

     async def some_task(args):
             ...

     def callback():
             ...
             newtask_queue.put(some_task(args))
              ...

This works because the `Queue.put()` method works from normal synchronous code.   The special `creator_task()` coroutine will pick it up and spawn the new task when control makes it back to the curio event loop. 

Note:  This part of curio is pretty experimental.  Let me know if this works... ;-).

**Comment by dabeaz** [comment]
Different comment.  Do you know how easy/hard it is to get the actual file descriptors being used for the Tk event loop?    If you could get access to one of those, you could probably have curio wait for I/O events.  Then call the Tk `update()` method as described.    That might be kind of neat.

**Comment by njsmith** [comment]
It looks like twisted's tkinter integration uses the same polling approach described in that reddit comment, so I'm guessing that's the best option available.

**Comment by nameoftherose** [comment]
I confirm that the Queue suggestion works.
I am a bit surprised though, as both the tutorial and the reference describe queue.put as an async function to be invoked with await.
Concerning the second suggestion, I am afraid I am no position to comment.

**Comment by dabeaz** [comment]
Was thinking about this more.  I bet you could simplify it further by simply putting all of the coroutines in a deque.  Then await them after you call the Tkinter update() method.  For example::

    pending = deque()

    def some_callback():
        ...
        pending.append(spawn(some_task())

    async def main():
         ...
         update()     # Tkinter update method
         while pending:
                await pending.popleft())

That's a bit rough, but I think it would work just like the Queue method.  Maybe even better.

**Comment by nameoftherose** [comment]
I think I see your point and it is not about deque.
But I still am of the opinion that no Queue or deque should be neccessary, there should be a way to to start the coroutine from the callback.

**Comment by dabeaz** [comment]
It seems that using queues are also a technique for communicating between Tkinter and threads.   

In addressing this, I've been doing a lot of thinking about the Curio Queue and Event objects.  Right now there are synchronous `put()` and `set()` methods.   However, maybe they are completely unneeded--actions that need to be awaited can be queued up in a deque and processed later.   This is much more general purpose and flexible.     So, I would encourage you to use that technique instead of the Queue object.  I'm probably going to remove the synchronous implementation of those methods (which were always highly experimental and not really documented).

**Comment by njsmith** [comment]
We definitely need some way to send work into the kernel from other
contexts -- including other threads -- that doesn't involve busy-polling.
(Busy polling is fine for the tkinter case since it requires a busy polling
loop anyway, but it's not a general solution.) Maybe
run_in_thread(threading.Queue.get) is that way, though, so curio.Queue
doesn't need to be involved? Though it would be good to have a primitive
that didn't require spending whole threads on just waiting around.

On Nov 19, 2016 09:18, "David Beazley" <notifications@github.com> wrote:

It seems that using queues are also a technique for communicating between
Tkinter and threads.

In addressing this, I've been doing a lot of thinking about the Curio Queue
and Event objects. Right now there are synchronous put() and set() methods.
However, maybe they are completely unneeded--actions that need to be
awaited can be queued up in a deque and processed later. This is much more
general purpose and flexible. So, I would encourage you to use that
technique instead of the Queue object. I'm probably going to remove the
synchronous implementation of those methods (which were always highly
experimental and not really documented).

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/dabeaz/curio/issues/111#issuecomment-261726501>, or mute
the thread

**Comment by dabeaz** [comment]
I think that queues should be the preferred mechanism for communicating with curio tasks from other contexts.    If threads are being used, use a thread Queue object and the curio `abide()` method as described here:  http://curio.readthedocs.io/en/latest/reference.html#synchronizing-with-threads-and-processes

Initially, I though that a queue could also be used to communicate between synchronous space and asynchronous space.  Now, I'm not so sure.  Yes, it can be done, but maybe it's enough to just use a deque or something simpler.   Coroutines are weird--you can create them externally and pass them into curio where they can be awaited later.  I need to think about that a bit more. 

I wasn't planning to give the low-level curio kernel itself any kind of API that allows it to be controlled from outside threads or non-async code.   Isolation is part of the whole design.  You interact with the kernel via async/await.  Always.

**Comment by njsmith** [comment]
Posted #119 to explain why `abide()` isn't sufficient, since that was getting off-topic here :-).

**Comment by dabeaz** [comment]
Just a note:  I don't think it's necessary to use a Queue and abide() for the case of Tkinter.  However, that approach would likely work fine if you're communicating between a single thread and a single curio task.   As noted in #119, you'd need to be careful if you were doing more advanced communication patterns such as have a pool of worker tasks waiting on a thread queue or something of that nature.

**Comment by swfiua** [comment]
Not tk, but if you want curio to run a qt event loop, the following does the trick:

async def qtloop(app):

    event_loop = qtcore.QEventLoop()

    while True:
        #for win in app.allWindows():
        #    win.show()
        event_loop.processEvents()
        app.sendPostedEvents(None, 0)

        # Experiment with sleep to keep gui responsive
        # but not a cpu hog.
        await curio.sleep(0.05)
    
curio.run(qtloop(app), with_monitor=True)

I was hoping to run qtconsole this way ... but it has threads, stinging bats and manta rays for now.

**Comment by dabeaz** [comment]
There's a part of me that thinks there MUST be a more sane way to to handle this than having curio periodically invoke the GUI event loop.   Maybe not.

**Comment by swfiua** [comment]
If qt were just pushing its events into a curio.Queue then that might do the trick.

**Comment by dabeaz** [comment]
It's really too bad curio can't I/O wait on the same "file" that's generating Qt/Tkinter events.  Curio doesn't actually perform the I/O so if it could do this, it could certainly transfer control to the GUI event polling loop when something did occur.

**Comment by swfiua** [comment]
That would be perfect.   Just need to find where that "file" is hidden.

In the case of pyqt there is the added fun that it is really just a wrapper
around the Qt C++ code, so we are at the mercy of whatever that does.

**Comment by dabeaz** [comment]
We just need to yank the file descriptor out of there using ctypes or something.  Then the swarm of stinging bats will have reached maximum density.

**Comment by njsmith** [comment]
The other potential strategy is to basically use the qt event loop as the
kernel's selector. Sounds annoying to set up, but has the advantage that in
principle it only needs regular qt APIs with all the hard work being on the
curio side.

**Comment by swfiua** [comment]
I'm back in tkinter land after an excursion to qt...

Just re-read this thread and the pieces are coming together.

I'm wanting to have a curio task managing the events from tk/qt/whatever.

I think the files involved can be found near here:

https://github.com/tcltk/tcl/blob/master/generic/tclMain.c#L662

Plus tkinter.Tk,foo() talks directly for a lot of foo's.

For now, I have a polling Tk loop that (will) do two things:

a. pass on events to the gui toolkit so it does its thing ie manage the screen

b. listen to input events:  keyboard, mouse, whatever..  and turn them into a common language.  That gets put into curio.UniversalQueue's

If we can avoid the polling then the UI performance should go way up.

From there lots of fascinating options.

Current goal: tickle swarms of glittering stinging bats 

Johnny

**Comment by imrn** [comment]
Can we have Kernel.pause/continue() functions?

**Comment by dabeaz** [comment]
What would they do?   In what context are they called?

**Comment by imrn** [comment]
Kernel.pause() pauses :) preserving all state.
Kernel.continue() resumes from where it left off.

It is for embedding the kernel into another application.
Parent application will provide pause, start, stop logic.
Coroutines may also decide to pause the kernel. 
(Of course they can't restart again.)

This enables to perform all operations within the thread of main
application. Communication between curio and parent app
will still be via queues, events, etc... Coroutines in curio can even
call sync functions of the parent app.

**Comment by imrn** [comment]
In fact, Kernel.pause() can only be called by coroutines.
This will cause Kernel.run() to return in the parent app.
Kernel.continue() will then be called by parent.

**Comment by imrn** [comment]
> Coroutines in curio can even call sync functions of the parent app.

A curio instance running in an other thread can too call functions
of the parent. However this brings unsafe thread behaviour.
Running parent and curio in the same thread is intrinsically
thread safe. But deadlocks are still possible. Developer
should be careful.

**Comment by imrn** [comment]
How can we call pause if we are going to have it?
Two example forms provided below. Should be have both sync and async?
I guess sync only option is sufficient since it can be called everywhere.
Ideas?

```python
from curio import pause, apause

async def going_to_pause():
    pause()
    await something()  # Before executing this kernel pauses. Kernel.run() returns.
    await apause()        # Same as above.
```

**Comment by dabeaz** [comment]
Is the purpose of this to somehow integrate Curio into a foreign event loop?   Can threads not be used instead?

**Comment by swfiua** [comment]
Penny just dropped.  I have a pathological fear of threads...  so try to pretend they don't exist.

So this should be pretty easy to do. 

For tkinter, just run all the tk stuff in a separate thread,

Have that do its thing, but communicate with curio via curio,queue's.

Now so long as the gui loop isn't polling (it's the polling I really want to get away from) then everything should just work.  It saves digging in the depths of the tcl C-code.

I am going to give this a go with tkinter and see what happens.

It will be something like.

        pig = curio.thread(build(meta()))

where meta is the meta data describing the gui, meta describes what to build.

Somewhere else there will be:

       event = await pig.event.pop()

       # do whatever you want with the event

I'll be experimenting over here:

   https://github.com/openbermuda/karmapi

I'll post a link if any of the commits do anything interesting.

**Comment by dabeaz** [comment]
On the whole, I'm not sure I'm so keen on building a whole bunch of pause/resume/continue functionality into Curio because before long it's likely to turn into a whole menagerie of "run_now", "run_soon", "run_at", "run_after", "run_before", "run_whenever",  "run_away", and "run_away_faster" functions.   

I suppose one could add some kind of option to the existing `run()` function that accepted a number of scheduling cycles to execute. Calling `run()` with no arguments currently resumes any tasks that are active.  Would need to think about it some more.

**Comment by imrn** [comment]
On 2/1/17, David Beazley <notifications@github.com> wrote:
> Is the purpose of this to somehow integrate Curio into a foreign event loop?

Yes it can be used that way and it can be used to to delegate some
functionality to curio in "existing" applications.

> Can threads not be used instead?

Yes, but virtues of single threaded application is also
great. Isn't ironic that a library targeting to reduce
dependency on threads recommends threading
as the only means of integration/embedding?

Any sane developer would use it at last resort.
Yes, I write this as I'm more comfortable with threads
than most, yet avoid them whenever possible.

The solution wouldn't be rewriting existing apps
in curio, or employing queues all over.. Not to
mention trying to visuliaze and avoid races.

You have above troubles on one hand, and
"Just calling parent app functions
from your coroutines without worrying about thread safety."
on the other. Which one is preferable?

Pushing for curio threads increases barrier of entry for
curio embedding. And enforces unnecessary complexity.

I got your point: For large scale we may not avoid threads.
But for majority of cases single threaded embedding is
the logical solution and will do a prety good job.

> On the whole, I'm not sure I'm so keen on building a whole bunch of
> pause/resume/continue functionality into Curio because before long it's
> likely to turn into a whole menagerie of "run_now", "run_soon", "run_at",
> "run_after", "run_before", "run_whenever",  "run_away", and
> "run_away_faster" functions.

Nobody wants it that fast..

> I suppose one could add some kind of option to the existing `run()` function
> that accepted a number of scheduling cycles to execute. Calling `run()` with
> no arguments currently resumes any tasks that are active.  Would need to
> think about it some more.

To sum up:
- Only pause() is needed. (If run() can cover continuing.)
- This will enable single threaded curio embedding. (Currently it cant be done)

**Comment by imrn** [comment]
> The solution wouldn't be rewriting existing apps
> in curio, or employing queues all over..

For some/most cases these are not even a possible options.

**Comment by imrn** [comment]
> I suppose one could add some kind of option to the existing 
> run() function that accepted a number of scheduling cycles to execute.

# of cycles may not be an appropiate measure for runs.
This is what operating systems doing to threads blindly.
That's why we are after cooperative scheduling.
Do you know a good framework doing it? :))

Anyway, lets think like this:
- run() as parent sync app yielding to curio.
- pause() as curio yielding to parent.

**Comment by dabeaz** [comment]
The thing that's a bit weird about `pause()` is how would code calling `run()` be able to distinguish between the completion of the supplied coroutine and pausing?   Is a special kind of return code made?  Is an exception thrown?    Sounds almost similar to the problem of a generator function itself--you'd need a different mechanism to handle the pause (yield) and the return (StopIteration).

**Comment by imrn** [comment]
Return code seems to be the simplest and most compatible approach for run(). 

Infact, I didn't mean any actual yielding or anything like iteration at the above post.
However, we can also think about it. What benefits can it provide? Will it add
additional state, context, location, etc bounded constraints (which we
would not like)? Ideas?

One thing I'm "almost" sure is that, pause() should be a thread safe (if it contains
thread sensitive code) sync function which can be called from anywhere. (From
coroutines and other threads.)

Note that caller of run() will be waiting on the line and curio is running coroutines
on the caller's thread, until deciding to give back control with pause(). Or until an other
thread decides that's enough.

**Comment by njsmith** [comment]
There are a number of possible approaches here. Most existing event loops
don't provide a "pause" method for this purpose, but instead provide a kind
of "pump" interface, where you can say "run for one iteration then return".
(Often with some flag to control whether the loop should block if nothing
is ready to run,  versus forcing the selector timeout to zero.) These
approaches are still pretty awkward because they rely on polling, though.
The next level of integration is to directly use the GUI event loop for I/o
(e.g. in Qt you can register sockets with the event loop and ask to be
notified when they're readable/writable). The downside here is that it
would require significant restructuring of the kernel, and I have no idea
what the public API would look like.

Right now I think the real question is, how do users *want* GUI + curio
code to work? What would be a pleasant and usable API? It's not at all
clear given how callback-happy existing toolkits are, but if we knew that
then we could start figuring out how to get there.

**Comment by dabeaz** [comment]
The `run()` method could probably be modified to do some kind of "pumping" in the sense of running through one scheduler cycle.  It would still require some thought though.

It might be theoretically possible to bridge Curio to a foreign event loop by making a custom ``Selector`` object of some kind (although I'm not completely certain).

**Comment by imrn** [comment]
I'll be happy to hear solid and better approaches for single threaded embedding
where host/guest app functions can call each other safely without the
need of awkward guards.

**Comment by imrn** [comment]
As return value of Kernel.run() has just been 'occupied' ;) by the coroutines
return value, we are left with others options.

I'm also negative about any other middleware (pumping, queue, etc)
between host/guest (or driver/driven, or gui/backend, etc) apps.

Because it undermines the very reason for putting all apps in a single
thread. It is: "Free form calls between apps".

The idea is: "Yield to the app and allow whatever it wants to do to others.
and let they exchange control as they see appropriate."

A plumber declaring itself as middleware is the last thing I would like to
see wandering around while the guest curio app making surgical hacking
into the guts of the host app.

**Comment by dabeaz** [comment]
I'll admit that I'm not so negative on threads.   However, if everything is going to live in a single thread, something has to be in control.   Is Curio in control or is it something else?   If it's something else, then I don't really know many other options besides pumping it (e.g., `run_once()` or whatever) or using callback functions (the latter not really being an option).

**Comment by imrn** [comment]
On 2/3/17, David Beazley <notifications@github.com> wrote:
> I'll admit that I'm not so negative on threads.

I'm not either. Even we can do it with
current toolbox. I don't mind enhancing
infrastructure for it.

But I'm saying another thing:
Allow full/deep embedding into the same thread space.

> However, if everything is
> going to live in a single thread, something has to be in control.
> Is Curio in control or is it something else?

In most cases the driver would be the existing sync host
app revised slightly to run/continue the kernel from different
locations of it code base. But there may be totally different layouts.

> If it's something else, then I don't
> really know many other options besides pumping it (e.g., `run_once()` or
> whatever) or using callback functions (the latter not really being an
> option).

Is there a rule that says there must predetermined number of event
loops, gui, sync or async app, masters, etc in a layout?

Besides why curio needs to know who is who?
It just needs to yield back to its caller which is determined by the logic
in the curio app. That's it..

- This makes it possible to put curio app (preserving its state) in
the same thread with others.
- And this in turn enables curio to call anything surrounds it.
Repeat: anything.
This should be something. Isn't it?

Anyway I find my self repeating all those. Please
let me know which one true. So I'll continue/discontinue
as it fits:

- Is it techinically too difficult/impossible to provide this?
- Or are we weighing possible options in pure tech ground?
- Or is it someting else..

**Comment by dabeaz** [comment]
Here is my general thought about how I might approach this.   The primary interface to curio is the `run()` method.   Right now, the `run()` method executes the supplied coroutine and all spawned non-daemonic tasks until they all exit.

I'm thinking I could probably relax the exit requirements.  Maybe I just make `run()` execute just until the supplied coroutine exits regardless of what other tasks might have been spawned.  If there were other tasks sitting in the background, they'd continue to stick around until `run()` is called again.   

It's not quite a `pause()` functionality, but whatever outside app is running could still submit a cororoutine to Curio to do something.  As soon as that coroutine returns, you're out of Curio and back to the app.

**Comment by imrn** [comment]
Well it can emulate pause() I guess:
- The pausable curio app can have pause Event() somewhere. (threading.Event if can be paused by external threads.)
- Parent coroutine submitted to run() should be just a wrapper for spawning other real tasks AND it waits for the pause Event() at the end.
- If any worker triggers the event, that's it: parent co exits and we are out of loop, and preparing for the next entry.
- Lets think twice so that modifications on run() doesn't interfere with current behaviour.

I guess it is done. Did I miss something? Rough pseudo code here:

```python
pause_evt = threading.Event()
# or just Event() if no other threads will pause us.

async pause_wrapper(co):
    await spawn(co)
    await abide(pause_evt.wait)
    # await pause_evt.wait()  # Or just this for asyncs only.
    pause_evt.clear()

def pausable_run(co):
   run(pause_wrapper(co))

co = co_that_can_stop_the_world(pause_evt)
pausable_run(co)
```

**Comment by imrn** [comment]
However, if we can have native curio pause() it would be more elegant.
From top of my head: Such as moving kernel queues to a temporary location
so that next run restores it. So that we prevent each run cycle to check for a pause.

**Comment by imrn** [comment]
Probably it isn't that easy but let me try:

This will throw us out of run() with Paused exception
leaving behind everything untouched.
Can we again reenter run() safely?

kernel.py:
```python
def pause():
    raise Paused

def run(self, coro=None, *, shutdown=False):
    # A lot of code here
    while njobs > 0:
        # and here
        for _ in range(len(ready)):
            current = ready_popleft()
            try:
                # some here
            # exceptions handlers here...
            # Now add this:
            except Paused:
                raise
```

**Comment by dabeaz** [comment]
If `Pause` is an exception, whatever task that raises it will simply terminate.    You won't be able to resume that task from that point forward (similar to what happens if a generator function raises an exception--the generator is done at that point). 

Regarding `run()`, it's already the case that calling `run()` with `None` is allowed.   There's really not a problem with calling `run()` multiple times.    It's just that the current behavior of `run()` is to execute until all non-daemonic tasks terminate.   The return value of `run()` is the return value of the coroutine initially provided to it.

The easiest change to make would be to have, as an option, something that makes `run()` return as soon as the initially supplied coroutine terminates instead of waiting around for everything else to terminate.   If you repeated called `run(None)` after that, whatever remaining tasks were still around would have a chance to run.   Or if you called `run(coro)` with a new coroutine, it would run until that coroutine terminated.

**Comment by njsmith** [comment]
> However, if we can have native curio pause() it would be more elegant

> The easiest change to make

"Elegant" and "easy" are both nice properties, but I still think "useful" should be in the discussion somewhere :-)

**Comment by imrn** [comment]
On 2/4/17, David Beazley <notifications@github.com> wrote:
> If `Pause` is an exception, whatever task that raises it will simply
> terminate.    You won't be able to resume that task from that point forward
> (similar to what happens if a generator function raises an exception--the
> generator is done at that point).

Did you noticed this above?

except Paused:
    # _cleanup_task(current)
    raise

Do we still lose that task?
Are there any way to make it continue on the next run?

> The easiest change to make would be to have, as an option, something that
> makes `run()` return as soon as the initially supplied coroutine terminates
> instead of waiting around for everything else to terminate.   If you
> repeated called `run(None)` after that, whatever remaining tasks were still
> around would have a chance to run.   Or if you called `run(coro)` with a new
> coroutine, it would run until that coroutine terminated.

Ok. First things first.
Can you provide it so that we can at least have the emulation?

**Comment by imrn** [comment]
On 2/4/17, Nathaniel J. Smith <notifications@github.com> wrote:
> "Elegant" and "easy" are both nice properties, but I still think "useful"
> should be in the discussion somewhere :-)

I'm already using threaded embedding of curio as you're all discussing above.

It is a very contraining setup such that all cross thread accesses
should go through some interface or they should be guarded.

I'll kick such trash out with and employ threads where they are
really needed.

**Comment by njsmith** [comment]
I totally agree that there is probably *some* nice way to integrate curio and GUI event loops. What I want to see is some sketch of what code would look like that uses these interfaces to accomplish something real. I can see how to implement this `pause` thing, but I can't see how to use it to actually write useful programs.

**Comment by imrn** [comment]
Ok let me be more direct then without providing food for demogogie:

- I need curio "Yielding" back to me. Fullstop.
- It's none of your business under what circumstances it will do that.
- I can make it going somehow.
- Either it stays private with me, or you too have on your repository.

**Comment by dabeaz** [comment]
I'm just trying to explore options of how this would work before looking at it further.   For example, that `Paused` exception approach won't work because if a task raises an exception and it propagates all the way out to the kernel, the task can no longer be resumed (because you can't resume any Python function at a point where an exception is raised).   That said,  I'm sure it's possible to do something.   Just not sure what the "something" will be yet.

**Comment by imrn** [comment]
Ok lets continue thinking about it. Here I have question:
Since it is generic thing, should I open a seperate issue for pause
or shall we continue here?

**Comment by dabeaz** [comment]
Yeah, maybe makes sense as a separate issue.

**Comment by dabeaz** [comment]
Have included two new examples in the examples directory with Tkinter/Curio interaction.
